rfd/0198-hardware-key-pin-caching.md

```diff
@@ -0,0 +1,189 @@
+
+```console
+> tsh proxy db --tunnel --port=60000
+Enter your YubiKey PIV PIN:
+Started authenticated tunnel for the PostgreSQL database "postgres" in cluster "root.example.com" on 127.0.0.1:60649.
+
+Use the following command to connect to the database or to the address above using other database GUI/CLI clients:
+  $ psql postgres://postgres@localhost:60000/postgres
+
+# User can open several postgres connections over the local proxy without
+# being re-prompted for pin, until the pin cache times out.
+#
+# Note that the user is not re-prompted for pin immediately after the cache times
+# out. Instead they are prompted the next time they try to create a psql connection
+# after the timeout.
+Enter your YubiKey PIV PIN:
+```
+
+### Multi Process Solution
+
+In order to cache the pin across multiple Teleport client processes (`tsh`,
+`tctl`, Teleport Connect), we can use the new [Teleport Key Agent](./0199-teleport-key-agent).
+
+Teleport Key Agent is responsible for providing an interface to signing with
+the user's private keys. When `hardware_key_pin` is required, the the user's
+PIV pin must be entered in order to perform a signature. Therefore, the agent
+is responsible for prompting and caching the pin.
+
+Any clients depending on the agent will share the benefits of the agent client's
+pin prompts and pin cache. For example, if one dependent client requests a
+signature and the user enters their pin into the agent, all dependent clients
+can continue with the cached pin for their signatures, until the pin cache
+times out.
+
+This solution will have far better UX when paired with Teleport Connect
+because Teleport Connect has the ability to pop into the foreground when the
+user needs to enter their PIV pin.
+
+### Security
+
+Caching the pin in process memory introduces a risk for the pin to be
+compromised. Fortunately, this risk can be almost entirely mitigated with
+secure memory practices implemented in [memguard](github.com/awnumar/memguard).
+
+Using [memguard](github.com/awnumar/memguard), Teleport clients will store the
+cached pin in a secure enclave.
+
+### Configuration
+
+#### Cluster Auth Preference
+
+To enable pin caching for Teleport clients, set `cap.hardware_key.pin_cache_timeout`
+to the desired timeout duration:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  ...
+  hardware_key:
+    # pin_cache_timeout is the amount of time that Teleport clients will cache
+    # the user's PIV pin. The timeout countdown is started when the pin is
+    # stored and is not extended by subsequent accesses.
+    pin_cache_timeout: 15m
+```
+
+Teleport Clients will retrieve this setting through `/webapi/ping`.
+
+### Proto
+
+```diff
+### types.proto
+message HardwareKey {
+  ...
++  // PinCacheTimeout is the amount of time in nanoseconds that Teleport clients
++  // will cache the user's PIV PIN when hardware key PIN policy is enabled.
++  int64 PinCacheTimeout = 3 [
```

Comment:
Call it "PinCacheTimeoutSeconds"?

Reply:
My bad, my brain read "seconds" somehow. +1 for adding whatever the time unit is to the field name.

Reply:
Fair enough!